File: core-js/zilingo/top-view-binary-tree.js

function in_order_traversal(root, distance_from_root, map, depth) {
    if (!root) return;
    if (!map.has(distance_from_root)) {
        map.set(distance_from_root, { data: root.data, depth });
    } else if (map.get(distance_from_root).depth > depth) {
        map.set(distance_from_root, { data: root.data, depth });
    }
    in_order_traversal(root.left, distance_from_root - 1, map, depth + 1);
    in_order_traversal(root.right, distance_from_root + 1, map, depth + 1);
}

/* 
Print top view of binary tree
Level order traversal + Vertical order traversal
Add hd (horizontal distance) for each node. For root node hd = 0
Do level order traversal
First node in level order traversal for a given hd comes in top view
*/
function top_view(root) {
    const queue = [];
    const node_distance_map = new Map();
    root.hd = 0;
    queue.push(root);
    while (queue.length) {
        const tmp = queue.shift();
        if (!node_distance_map.has(tmp.hd)) {
            node_distance_map.set(tmp.hd, tmp.data);
        }
        if (tmp.left) {
            tmp.left.hd = tmp.hd - 1;
            queue.push(tmp.left);
        }
        if (tmp.right) {
            tmp.right.hd = tmp.hd + 1;
            queue.push(tmp.right);
        }
    }
    console.log(node_distance_map);
    return node_distance_map;
}
function main() {
    let root = {
        data: 1,
        left: {
            data: 2,
            left: {
                data: 4,
                left: null,
                right: null
            },
            right: {
                data: 5,
                left: null,
                right: {
                    data: 7,
                    left: null,
                    right: {
                        data: 8,
                        left: null,
                        right: {
                            data: 9,
                            left: null,
                            right: null
                        }
                    }
                }
            }
        },
        right: {
            data: 3,
            left: null,
            right: {
                data: 6,
                left: null,
                right: null
            }
        }
    }
    // const node_distance_map = level_order_traversal(root);
    // console.log(node_distance_map.keys());
    // for (let key of Array.from(node_distance_map.keys()).sort((a, b) => a - b)) {
    //     console.log(node_distance_map.get(key));
    // }
    const map = new Map();
    in_order_traversal(root, 0, map, 0);
    const top_view_arr = [];
    for (let [key, value] of map) {
        top_view_arr.push(value);
    }
    console.log(top_view_arr);
}
main();

File: core-js/zilingo/top-view-binary-tree.py
# https://www.hackerrank.com/challenges/tree-top-view/copy-from/14110629

class Node:
    def __init__(self, val, level):
        self.right = None
        self.info = val
        self.left = None
        self.level = level

def top_view(root):
    queue = []
    node_distance_dict = dict()
    root.hd = 0
    queue.append(root)
    while queue:
        tmp = queue.pop(0)
        if not node_distance_dict.has_key(tmp.hd):
            node_distance_dict[tmp.hd] = tmp
        if tmp.left:
            tmp.left.hd = tmp.hd - 1
            queue.append(tmp.left)
        if tmp.right: 
            tmp.right.hd = tmp.hd + 1
            queue.append(tmp.right)
    return node_distance_dict
    

def main():
    root = Node(1, 0)
    root.left = Node(2, 1)
    root.right = Node(3, 1)
    root.left.left = Node(4, 2)
    root.left.right = Node(5, 2)
    root.right.right = Node(6, 2)
    root.left.right.right = Node(7, 3)
    root.left.right.right.right = Node(8, 4)
    root.left.right.right.right.right = Node(9, 5)
    # print(root.right.right.data)
    info_dict = top_view(root)
    arr = []
    for distance in info_dict:
        arr.append((info_dict[distance].info, distance))
    arr_sorted = sorted(arr, key=lambda n: n[1])
    for i in range(0, len(arr_sorted)):
        print arr_sorted[i][0],
    

main()

